TestNG parallel test execution via Gradle fails with "Received a completed event for test with unknown id".

<!--- 
Please follow the instructions below. We receive dozens of issues every week, so to stay productive, we will close issues that don't provide enough information. 

Please open Android-related issues on the Android Issue Tracker at https://source.android.com/source/report-bugs
Please open Kotlin DSL-related issues at https://github.com/gradle/kotlin-dsl/issues
Please open Gradle Native-related issues at https://github.com/gradle/gradle-native/issues
-->

<!--- Provide a brief summary of the issue in the title above -->
The appears to relate to https://github.com/gradle/gradle/issues/6453, but for TestNG instead of JUnit.  I attempted some of the fixes described in that ticket (turn off parallel execution, use the testng serialized annotation), but nothing worked.  I ran into this bug on the 5.0 release, as well as a recent 5.1 build (gradle-5.1-20181128000036+0000-bin), so I do not think it is yet fixed.

### Expected Behavior
All tests are executed normally
<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->

### Current Behavior
Tests fail.

Stack:
`Received a completed event for test with unknown id '1.301'. Registered test ids: '[1.1, 1.33, 1.3, 1.35, :modules-offheap:test, 1.2, 1.229, 1.232, 1.31]'
java.lang.IllegalArgumentException: Received a completed event for test with unknown id '1.301'. Registered test ids: '[1.1, 1.33, 1.3, 1.35, :modules-offheap:test, 1.2, 1.229, 1.232, 1.31]'
        at org.gradle.api.internal.tasks.testing.results.StateTrackingTestResultProcessor.completed(StateTrackingTestResultProcessor.java:79)
        at org.gradle.api.internal.tasks.testing.results.AttachParentTestResultProcessor.completed(AttachParentTestResultProcessor.java:56)
        at jdk.internal.reflect.GeneratedMethodAccessor578.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.dispatch.FailureHandlingDispatch.dispatch(FailureHandlingDispatch.java:29)
        at org.gradle.internal.dispatch.AsyncDispatch.dispatchMessages(AsyncDispatch.java:87)
        at org.gradle.internal.dispatch.AsyncDispatch.access$000(AsyncDispatch.java:36)
        at org.gradle.internal.dispatch.AsyncDispatch$1.run(AsyncDispatch.java:71)
        at org.gradle.internal.concurrent.InterruptibleRunnable.run(InterruptibleRunnable.java:42)
        at org.gradle.internal.operations.CurrentBuildOperationPreservingRunnable.run(CurrentBuildOperationPreservingRunnable.java:42)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.base/java.lang.Thread.run(Thread.java:834)`


<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->

### Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->
I have been attempting to upgrade to Gradle 5 in order to switch to java 11 (see https://github.com/gradle/gradle/issues/7835) since there are no plans to backport ASM fixes into Gradle 4.10.  I was able to get past the other compatibility problems and got stymied by this bug.  I have disabled the offending tests for the time being, but I would like to turn them back on.

### Steps to Reproduce (for bugs)
<!--- Provide a self-contained example project (as an attached archive or a Github project). -->
<!--- In the rare cases where this is infeasible, we will also accept a detailed set of instructions. -->
I have not extracted a small project to reproduce this yet, but will do so if requested.  The test in question uses a "@DataProvider" and executes hundreds of very short running tests.

### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan `https://scans.gradle.com/get-started` is ideal -->
OS	CentOS Linux release 7.4.1708
CPU cores	8 cores
Java version	openjdk version "11.0.1" 2018-10-16

`> Task :test FAILED
Received a completed event for test with unknown id '1.9'. Registered test ids: '[1.1, 1.3, 1.2, 1.4, :test]'
java.lang.IllegalArgumentException: Received a completed event for test with unknown id '1.9'. Registered test ids: '[1.1, 1.3, 1.2, 1.4, :test]'
        at org.gradle.api.internal.tasks.testing.results.StateTrackingTestResultProcessor.completed(StateTrackingTestResultProcessor.java:79)
        at org.gradle.api.internal.tasks.testing.results.AttachParentTestResultProcessor.completed(AttachParentTestResultProcessor.java:56)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.dispatch.FailureHandlingDispatch.dispatch(FailureHandlingDispatch.java:29)
        at org.gradle.internal.dispatch.AsyncDispatch.dispatchMessages(AsyncDispatch.java:87)
        at org.gradle.internal.dispatch.AsyncDispatch.access$000(AsyncDispatch.java:36)
        at org.gradle.internal.dispatch.AsyncDispatch$1.run(AsyncDispatch.java:71)
        at org.gradle.internal.concurrent.InterruptibleRunnable.run(InterruptibleRunnable.java:42)
        at org.gradle.internal.operations.CurrentBuildOperationPreservingRunnable.run(CurrentBuildOperationPreservingRunnable.java:42)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.base/java.lang.Thread.run(Thread.java:834)

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> java.lang.NullPointerException (no error message)`

--- a/subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/testng/TestNGIntegrationTest.groovy
+++ b/subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/testng/TestNGIntegrationTest.groovy
@@ -276,6 +276,42 @@ class TestNGIntegrationTest extends MultiVersionIntegrationSpec {
             .assertTestFailed("failed to execute tests", containsString("Could not execute test class 'com.example.Foo'"))
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/7878")
+    def "can concurrently execute the same test class multiple times"() {
+        given:
+        file('src/test/java/TestNG7878.java') << """
+            import org.testng.annotations.Factory;
+            import org.testng.annotations.Test;
+            import org.testng.Assert;
+            
+            public class TestNG7878 {
+                @Factory
+                public static Object[] createTests() {
+                    return new Object[]{ 
+                            new TestNG7878(), 
+                            new TestNG7878() 
+                    };
+                }
+            
+                @Test
+                public void runFirst() {}
+                
+                @Test(dependsOnMethods = "runFirst")
+                public void testGet2() {
+                    Assert.assertEquals(true, true);
+                }
+            }
+        """
+
+        when:
+        succeeds('test')
+
+        then:
+        new DefaultTestExecutionResult(testDirectory)
+            .assertTestClassesExecuted('TestNG7878')
+            .testClass('TestNG7878').assertTestCount(4, 0, 0)
+    }
+
     private static String testListener() {
         return '''
             def listener = new TestListenerImpl()

--- a/subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/testng/TestNGTestResultProcessorAdapter.java
+++ b/subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/testng/TestNGTestResultProcessorAdapter.java
@@ -49,7 +49,7 @@ public class TestNGTestResultProcessorAdapter implements ISuiteListener, ITestLi
     private final Map<ITestContext, Object> testId = new HashMap<ITestContext, Object>();
     private final Map<ISuite, Object> suiteId = new HashMap<ISuite, Object>();
     private final Map<XmlTest, Object> xmlTestIds = new HashMap<XmlTest, Object>();
-    private final Map<ITestClass, Object> testClassId = new HashMap<ITestClass, Object>();
+    private final Map<ITestClass, TestClassInfo> testClassInfo = new HashMap<ITestClass, TestClassInfo>();
     private final Map<ITestResult, Object> testMethodId = new HashMap<ITestResult, Object>();
     private final Map<ITestNGMethod, Object> testMethodParentId = new HashMap<ITestNGMethod, Object>();
     private final Set<ITestResult> failedConfigurations = new HashSet<ITestResult>();
@@ -91,17 +91,24 @@ public class TestNGTestResultProcessorAdapter implements ISuiteListener, ITestLi
 
     @Override
     public void onBeforeClass(ITestClass testClass) {
-        TestDescriptorInternal testInternal;
-        Object parentId;
+        TestDescriptorInternal testInternal = null;
+        Object parentId = null;
         synchronized (lock) {
-            testInternal = new DefaultTestClassDescriptor(idGenerator.generateId(), testClass.getName());
-            testClassId.put(testClass, testInternal.getId());
-            parentId = xmlTestIds.get(testClass.getXmlTest());
-            for (ITestNGMethod method : testClass.getTestMethods()) {
-                testMethodParentId.put(method, testInternal.getId());
+            TestClassInfo info = testClassInfo.get(testClass);
+            if (info != null) {
+                info.incrementStartedCount();
+            } else {
+                testInternal = new DefaultTestClassDescriptor(idGenerator.generateId(), testClass.getName());
+                testClassInfo.put(testClass, new TestClassInfo(testInternal.getId()));
+                parentId = xmlTestIds.get(testClass.getXmlTest());
+                for (ITestNGMethod method : testClass.getTestMethods()) {
+                    testMethodParentId.put(method, testInternal.getId());
+                }
             }
         }
-        resultProcessor.started(testInternal, new TestStartEvent(clock.getCurrentTime(), parentId));
+        if (testInternal != null) {
+            resultProcessor.started(testInternal, new TestStartEvent(clock.getCurrentTime(), parentId));
+        }
     }
 
     @Override
@@ -110,9 +117,16 @@ public class TestNGTestResultProcessorAdapter implements ISuiteListener, ITestLi
 
     @Override
     public void onAfterClass(ITestClass testClass) {
-        Object id;
+        Object id = null;
         synchronized (lock) {
-            id = testClassId.remove(testClass);
+            TestClassInfo info = testClassInfo.get(testClass);
+            if (info != null) {
+                if (info.startedCount == 1) {
+                    id = testClassInfo.remove(testClass).id;
+                } else {
+                    info.decrementStartedCount();
+                }
+            }
         }
         // Guard against TestNG calling this hook more than once with the same testClass.
         // See https://github.com/cbeust/testng/issues/1618 for details.
@@ -263,22 +277,43 @@ public class TestNGTestResultProcessorAdapter implements ISuiteListener, ITestLi
 
     @Override
     public void onConfigurationFailure(ITestResult testResult) {
+        ITestNGMethod testMethod = testResult.getMethod();
+        ITestClass testClass = testMethod.getTestClass();
+        TestClassInfo classInfo;
         synchronized (lock) {
             if (!failedConfigurations.add(testResult)) {
                 // workaround for bug in TestNG 6.2 (apparently fixed in some 6.3.x): listener is notified twice per event
                 return;
             }
+            classInfo = this.testClassInfo.get(testClass);
         }
         // Synthesise a test for the broken configuration method
-        ITestNGMethod testMethod = testResult.getMethod();
-        ITestClass testClass = testMethod.getTestClass();
         TestDescriptorInternal test = new DefaultTestMethodDescriptor(idGenerator.generateId(), testClass.getName(), testMethod.getMethodName());
-        resultProcessor.started(test, new TestStartEvent(testResult.getStartMillis(), testClassId.get(testClass)));
+        Object parentId = classInfo == null ? null : classInfo.id;
+        resultProcessor.started(test, new TestStartEvent(testResult.getStartMillis(), parentId));
         resultProcessor.failure(test.getId(), testResult.getThrowable());
         resultProcessor.completed(test.getId(), new TestCompleteEvent(testResult.getEndMillis(), TestResult.ResultType.FAILURE));
     }
 
     @Override
     public void beforeConfiguration(ITestResult tr) {
+    }
+
+    static class TestClassInfo {
+
+        private final Object id;
+        private int startedCount = 1;
+
+        TestClassInfo(Object id) {
+            this.id = id;
+        }
+
+        public void incrementStartedCount() {
+            startedCount++;
+        }
+
+        public void decrementStartedCount() {
+            startedCount--;
+        }
     }
 }

